pkg/server/kubeconfig.go

```diff
@@ -0,0 +1,162 @@
+	tokenHashFileName := filepath.Join(rootDirectory, adminTokenStore)
+	adminTokenHash, err = ioutil.ReadFile(tokenHashFileName)
+	if os.IsNotExist(err) {
+		newAdminToken = uuid.New().String()
+		sum := sha256.Sum256([]byte(newAdminToken))
+		adminTokenHash = sum[:]
+		if err := ioutil.WriteFile(tokenHashFileName, adminTokenHash, 0600); err != nil {
+			return "", nil, err
+		}
+	}
+
+	if authn == nil {
+		// prevent nil pointer panic
+		return
+	}
+	if authn.Authenticator == nil {
+		// authenticator or authorizer might be nil if we want to bypass authz/authn
+		// and we also do nothing in this case.
+		return
+	}
+
+	var uid = uuid.New().String()
+	adminUser := &user.DefaultInfo{
+		Name:   user.APIServerUser,
+		UID:    uid,
+		Groups: []string{user.SystemPrivilegedGroup},
+	}
+
+	newAuthenticator := bearertoken.New(authenticator.WrapAudienceAgnosticToken(authn.APIAudiences, authenticator.TokenFunc(func(ctx context.Context, requestToken string) (*authenticator.Response, bool, error) {
+		requestTokenHash := sha256.Sum256([]byte(requestToken))
+		if !bytes.Equal(requestTokenHash[:], adminTokenHash) {
+			return nil, false, nil
+		}
+		return &authenticator.Response{User: adminUser}, true, nil
+
+	})))
+
+	authn.Authenticator = authenticatorunion.New(newAuthenticator, authn.Authenticator)
+
+	return newAdminToken, adminTokenHash, nil
+}
+
+func createKubeConfig(adminUserName, adminBearerToken, baseHost, tlsServerName string, caData []byte) *clientcmdapi.Config {
+	var kubeConfig clientcmdapi.Config
+	//Create Client and Shared
+	kubeConfig.AuthInfos = map[string]*clientcmdapi.AuthInfo{
+		adminUserName: {Token: adminBearerToken},
+	}
+	kubeConfig.Clusters = map[string]*clientcmdapi.Cluster{
+		// cross-cluster is the virtual cluster running by default
+		"cross-cluster": {
+			Server:                   baseHost + "/clusters/*",
+			CertificateAuthorityData: caData,
+			TLSServerName:            tlsServerName,
+		},
+		// admin is the virtual cluster running by default
+		"admin": {
+			Server:                   baseHost,
+			CertificateAuthorityData: caData,
+			TLSServerName:            tlsServerName,
+		},
+		// user is a virtual cluster that is lazily instantiated
+		"user": {
+			Server:                   baseHost + "/clusters/user",
+			CertificateAuthorityData: caData,
+			TLSServerName:            tlsServerName,
+		},
+	}
+	kubeConfig.Contexts = map[string]*clientcmdapi.Context{
+		"cross-cluster": {Cluster: "cross-cluster", AuthInfo: adminUserName},
+		"admin":         {Cluster: "admin", AuthInfo: adminUserName},
+		"user":          {Cluster: "user", AuthInfo: adminUserName},
+	}
+	kubeConfig.CurrentContext = "admin"
+	return &kubeConfig
+}
+
+func createKubeConfigs(server *server.GenericAPIServer, externalKubeConfigAdminToken string, externalKubeConfigAdminTokenHash []byte, externalKubeConfigPath string) (loopbackBasedKubeConfig *clientcmdapi.Config, externalKubeConfig *clientcmdapi.Config, err error) {
```

Comment:
please split into one for loopback, one for admin

Reply:
Fix in commit 7c3486cb3506a17f776e0a8fc6bcf5376c42d947